### ISSUE TITLE
Alternative to webpack externals in fuse-box?

### ISSUE THREAD

**Issue by direct-fuel-injection** [fuse-box/fuse-box#89]
How can I do this?
```
module.exports = {
...
  externals: {
    'react': 'React',
    'react-dom': 'ReactDOM'
  },
...
};
```

**Comment by nchanged** [comment]
Yeah check "globals" section on the doc server (fuse-box.org)

**Comment by nchanged** [comment]
[here](http://fuse-box.org/#global-variables)

**Comment by direct-fuel-injection** [comment]
```
var fuse = new fusebox.FuseBox({
    homeDir: buildDir,
    outFile: path.join(buildDir, 'bundle.js'),
    globals: {
        react: 'React',
        'react-dom': 'reactDom'
    },
    plugins: [
        fusebox.EnvPlugin({
            'NODE_ENV': JSON.stringify('production')
        }),
        fusebox.CSSPlugin(),
        fusebox.BabelPlugin({
            config: {
                sourceMaps: true,
                presets: ['es2015', 'react', 'stage-0']
            }
        })
    ]
});

...

gulp.task('build', function () {
    return fuse.bundle('>index.js -react -react-dom');
});

```
I have an error:
**`bundle.js:16078 Uncaught Package was not found "react-dom"`**

**Comment by nchanged** [comment]
You are not importing it, either do that or add +react-dom to your bundle

**Comment by nchanged** [comment]
Regardless, check the react example

**Comment by direct-fuel-injection** [comment]
@nchanged after I've added +react-dom, it's included in my bundle. 200kb -> 700 kb

**Comment by nchanged** [comment]
Well, it is trying to require it, and can't find it. Of course, if you want to separate them, you need to bundle react-dom with fuse first `+react-dom` and then include it as a script tag

**Comment by direct-fuel-injection** [comment]
@nchanged 

```
gulp.task('build', function () {
    return fuse.bundle('>index.js +react-dom');
});
```
react-dom added to bundle

**Comment by nchanged** [comment]
drop `index.js` from there, bundle `react-dom` separetely

**Comment by direct-fuel-injection** [comment]
@nchanged can you write task for gulp? can't handle it.

**Comment by nchanged** [comment]
Sure, a bit later!

**Comment by nchanged** [comment]
```
FuseBox.init({
    homeDir: "src",
    outFile: "react.dom.js",
}).bundle("+react-dom");
```

**Comment by direct-fuel-injection** [comment]
@nchanged where I should place it?

**Comment by direct-fuel-injection** [comment]
@nchanged where should I place it?
```
gulp.task('build', function () {
    return fusebox.init({
            homeDir: "src",
            outFile: "react.dom.js"
        }).bundle("+react-dom");
});
```
I have error on gulp build: `TypeError: fusebox.init is not a function`

**Comment by nchanged** [comment]
Sorry `FuseBox.init` create a separate task to build it, (you can do it only once) and then add it as a script tag

**Comment by direct-fuel-injection** [comment]
Why to bundle react-dom? Have already minimized.

**Comment by direct-fuel-injection** [comment]
Config:
```
var fuseboxConfig = {
    homeDir: buildDir,
    outFile: path.join(buildDir, 'bundle.js'),
    plugins: [
        fusebox.EnvPlugin({
            'NODE_ENV': JSON.stringify('production')
        }),
        fusebox.CSSPlugin(),
        fusebox.BabelPlugin({
            config: {
                sourceMaps: true,
                presets: ['es2015', 'react', 'stage-0']
            }
        })
    ]
};
```

Now I have two tasks:
```
gulp.task('reactdom', function () {
    return fusebox.FuseBox.init({
        homeDir: buildDir,
        outFile: path.join(buildDir, 'react.dom.js')
    }).bundle('+react +react-dom');
});

gulp.task('build', function () {
    return fuse.bundle('>index.js -react -react-dom');
});
```

My index.html:
```
<script src="react.dom.js"></script>
<script src="bundle.js"></script>
```
Build log:

```
Starting 'reactdom'...
└── default (0 files,  53 Bytes)
└── react (25 files) 119.3 kB
└── object-assign (1 files) 2.1 kB
└── fbjs (23 files) 33.2 kB
└── process (1 files) 3 kB
└── react-dom (127 files) 549.1 kB

Starting 'build'...
└── default (18 files,  44.9 kB)
└── babel-polyfill (1 files) 1 kB
└── core-js (294 files) 230.5 kB
└── regenerator-runtime (1 files) 22.2 kB
└── process (1 files) 3 kB
└── react-redux (15 files) 40.5 kB
└── react (25 files) 119.3 kB
└── object-assign (1 files) 2.1 kB
└── fbjs (4 files) 5.7 kB
└── hoist-non-react-statics (1 files) 1.6 kB
└── invariant (1 files) 1.7 kB
└── redux (7 files) 23.2 kB
└── lodash (10 files) 6.8 kB
└── symbol-observable (3 files) 1.5 kB
└── seamless-immutable (1 files) 23.3 kB
└── redux-thunk (1 files) 728 Bytes
└── redux-batched-actions (1 files) 737 Bytes
└── deep-diff (1 files) 11.4 kB

```
That's not great to bundle react-dom. React still in main module.
This works, but you say above that i need to use `globals`?

**Comment by nchanged** [comment]
if you need to expose it to window (unless react is already exposing it)

You can try isolating your project dependencies like so:

`fuse.bundle('>[index.js]');` so only your project files will be bundled

**Comment by direct-fuel-injection** [comment]
The problem is when fuse-box creates bundle without react-dom package, globals doesn't work properly. It's still trying to require package, not using global.

**Comment by nchanged** [comment]
You can make them [global](http://fuse-box.org/#global-variables)

For example:

```
globals: { lodash: "_", react : "react", "react-dom: "reactDom" }
```

**Comment by direct-fuel-injection** [comment]
Externals configuration in webpack provides a way of not including a dependency in the bundle. Instead the created bundle relies on that dependency to be present in the consumers environment.
Globals working not like this.

**Comment by nchanged** [comment]
@direct-fuel-injection we don't have it in place, but there are workaround that you can apply.

**Comment by nchanged** [comment]
Ok, I am re-opening the issue. Things are clear now. This feature is coming soon.

**Comment by nchanged** [comment]
@direct-fuel-injection please check this [section](http://fuse-box.org/#shimming)

**Comment by direct-fuel-injection** [comment]
@nchanged nice work here.
Config:

```
var fuseboxConfig = {
    homeDir: buildDir,
    outFile: path.join(buildDir, 'bundle.js'),
    cache: false,
    shim: {
        'react': {
            exports: 'React'
        },
        'react-dom': {
            exports: 'ReactDOM'
        }
    },
    plugins: [
        fusebox.EnvPlugin({
            'NODE_ENV': JSON.stringify('production')
        }),
        fusebox.CSSPlugin(),
        fusebox.BabelPlugin({
            config: {
                sourceMaps: true,
                presets: ['es2015', 'react', 'stage-0']
            }
        }),
        fusebox.UglifyJSPlugin({
            compress: { warnings: false },
            comments: false,
            mangle: true,
            minimize: true
        })
    ]
};
```

Build log:
```
Starting 'build'...
└── default (22 files,  48.3 kB)
└── react (1 files) 207 Bytes
└── react-dom (1 files) 214 Bytes
└── babel-polyfill (1 files) 1 kB
└── core-js (294 files) 230.5 kB
└── regenerator-runtime (1 files) 22.2 kB
└── process (1 files) 3 kB
└── react-redux (15 files) 40.5 kB
└── hoist-non-react-statics (1 files) 1.6 kB
└── invariant (1 files) 1.7 kB
└── redux (7 files) 23.2 kB
└── lodash (10 files) 6.8 kB
└── symbol-observable (3 files) 1.5 kB
└── seamless-immutable (1 files) 23.3 kB
└── redux-thunk (1 files) 728 Bytes
└── redux-batched-actions (1 files) 737 Bytes
└── deep-diff (1 files) 11.4 kB

    --------------
    Size: 417 kB 
    Time: 3s 928ms

```
But my bundle.js file size: `184,6 kB` (after uglify). Is it right to not show result size?
And I have` Warning: bundle.js:6 You are currently using minified code outside of NODE_ENV === 'production'. This means that you are running a slower development build of Redux. You can use loose-envify (https://github.com/zertosh/loose-envify) for browserify or DefinePlugin for webpack (http://stackoverflow.com/questions/30030031) to ensure you have the correct code for your production build.` EnvPlugin define NODE_ENV to 'production'.

**Comment by nchanged** [comment]
@direct-fuel-injection that's probably a bug in Uglify plugin, which is not affecting the log ... 

Will work just fine >>
```
fusebox.EnvPlugin({
      NODE_ENV: "production"
}),
```

Regarding the warning, who is the initiator? react?

**Comment by direct-fuel-injection** [comment]
@nchanged disabling stringify works thank you no warnings now.

**Comment by nchanged** [comment]
Great, now it's resolved. Thanks for submitting the bug. And join the gitter channel, if you want to discuss something.
